Promote PodDisruptionBudget to policy/v1

mortent opened the pull request:
#### What type of PR is this?
/kind api-change
/kind feature

#### What this PR does / why we need it:
This promotes `PodDisruptionBudget` resource to `policy/v1` based on https://github.com/kubernetes/enhancements/issues/85 and detailed description in https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/85-Graduate-PDB-to-Stable

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/95083

#### Special notes for your reviewer:
This replaces an older PR at https://github.com/kubernetes/kubernetes/pull/81571. 

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
The PodDisruptionBudget API has been promoted to policy/v1 with no schema changes. The only functional change is that an empty selector (`{}`) written to a policy/v1 PodDisruptionBudget now selects all pods in the namespace. The behavior of the policy/v1beta1 API remains unchanged. The policy/v1beta1 PodDisruptionBudget API is deprecated and will no longer be served in 1.25+.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/85-Graduate-PDB-to-Stable
```

mortent committed:
Promote PDBs to GA

soltysh commented on pkg/apis/policy/install/install.go:
```diff
@@ -34,5 +35,6 @@ func init() {
 	utilruntime.Must(v1beta1.AddToScheme(scheme))
-	utilruntime.Must(scheme.SetVersionPriority(v1beta1.SchemeGroupVersion))
+	utilruntime.Must(v1.AddToScheme(scheme))
+	utilruntime.Must(scheme.SetVersionPriority(v1beta1.SchemeGroupVersion, v1.SchemeGroupVersion))
```
v1 should always be before v1beta1

mortent replied:
So it seems like we should prefer v1beta1 now and then switch in the next release. Ref: https://github.com/kubernetes/kubernetes/pull/95718#discussion_r520969477

soltysh replied:
Fair, we'll switch in the next release. Add a todo here.

mortent replied:
Added a TODO here.

soltysh commented on pkg/features/kube_features.go:
```diff
@@ -735,7 +735,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 	StartupProbe:                                   {Default: true, PreRelease: featuregate.GA, LockToDefault: true}, // remove in 1.23
 	AllowInsecureBackendProxy:                      {Default: true, PreRelease: featuregate.Beta},
-	PodDisruptionBudget:                            {Default: true, PreRelease: featuregate.Beta},
+	PodDisruptionBudget:                            {Default: true, PreRelease: featuregate.GA, LockToDefault: true},
```
I'd consider this bit in a separate commit.

mortent replied:
I have removed this from the PR.

soltysh commented on staging/src/k8s.io/api/policy/v1/types.go:
```diff
@@ -0,0 +1,110 @@
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:prerelease-lifecycle-gen:introduced=1.5
+// +k8s:prerelease-lifecycle-gen:deprecated=1.22
```
Drop these, they are not needed.

janetkuo replied:
Good catch

mortent replied:
I get errors when I run the code gen if I remove these. Do you know if I need to make other changes to fix it? I'm still trying to understand why this happens.

soltysh commented on staging/src/k8s.io/api/policy/v1/types.go:
```diff
@@ -0,0 +1,110 @@
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:prerelease-lifecycle-gen:introduced=1.5
+// +k8s:prerelease-lifecycle-gen:deprecated=1.22
```
Ditto.

soltysh commented on staging/src/k8s.io/api/policy/v1beta1/types.go:
```diff
@@ -84,6 +84,8 @@ type PodDisruptionBudgetStatus struct {
 // +k8s:prerelease-lifecycle-gen:deprecated=1.22
 
+// DEPRECATED - This group version of PodDisruptionBudget is deprecated by policy/v1/PodDisruptionBudget. See
+// the release notes for more information.
```
Not needed, this is handled by the above tags automatically.

mortent replied:
Removed

soltysh reviewed (changes requested):
I left you some initial comments, you're also missing necessary changes to `pkg/api/testing/defaulting_test.go` which you can see from unit-test failure from that package.

fedebongio commented:
/cc @wojtek-t 
/remove-sig api-machinery

janetkuo commented on staging/src/k8s.io/api/policy/v1beta1/types.go:
```diff
@@ -84,6 +84,8 @@ type PodDisruptionBudgetStatus struct {
 // +k8s:prerelease-lifecycle-gen:introduced=1.5
 // +k8s:prerelease-lifecycle-gen:deprecated=1.22
```
You can use this to indicate which API users should use instead 

```
// +k8s:prerelease-lifecycle-gen:replacement=policy,v1,PodDisruptionBudget
```

mortent replied:
Added.

janetkuo commented on staging/src/k8s.io/api/policy/v1beta1/types.go:
```diff
@@ -84,6 +84,8 @@ type PodDisruptionBudgetStatus struct {
 // +k8s:prerelease-lifecycle-gen:introduced=1.5
 // +k8s:prerelease-lifecycle-gen:deprecated=1.22
```
Should we announce deprecation in 1.21 since policy/v1. PodDisruptionBudget would be available?

mortent commented:
/test pull-kubernetes-integration

liggitt commented on staging/src/k8s.io/api/policy/v1/types.go:
```diff
@@ -0,0 +1,110 @@
+	// +optional
+	MinAvailable *intstr.IntOrString `json:"minAvailable,omitempty" protobuf:"bytes,1,opt,name=minAvailable"`
+
+	// Label query over pods whose evictions are managed by the disruption
+	// budget.
```
I don't see the selector change proposed in https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/85-Graduate-PDB-to-Stable/README.md#fix-handling-of-empty-selector-in-disruption-controller implemented here

mortent replied:
Yeah, I haven't done this yet. Just wanted to get the basics in place first. This is next on the list.

mortent replied:
I have updated the PR with an attempt to address this, but I'm not sure it is correct. So my assumptions are:
 * We can't change the version of PDB used in the controller yet, since we want to update the API first and then only update the controller in the next release.
 * There will existing PDBs stored in etcd, so we also can't change the semantics of an empty selector in the internal type.

The result is that the conversion must happen in the v1 api as we convert between v1 and the internal type.

liggitt replied:
> There will existing PDBs stored in etcd, so we also can't change the semantics of an empty selector in the internal type.

etcd stores in v1beta1, not internal. I would expect to make adjustments in the `v1beta1 <-> internal` conversion

liggitt replied:
> We can't change the version of PDB used in the controller yet, since we want to update the API first and then only update the controller in the next release.

if you want to add a conformance test, the controller must use the v1 API (conformance tests must pass in clusters with no alpha/beta features or APIs enabled)

mortent replied:
That makes sense. I have updated the PR to handle conversion in v1beta1 and updated the controller to use the v1 api.

mortent commented on staging/src/k8s.io/api/policy/v1/types.go:
```diff
@@ -0,0 +1,110 @@
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	Items           []PodDisruptionBudget `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
```
So can we update the version of Eviction since the v1beta1 is part of the pod API (which is GA)? We had a discussion around this in my previous PR to take PDB to GA: https://github.com/kubernetes/kubernetes/pull/81571#discussion_r318874136 and https://github.com/kubernetes/kubernetes/pull/81571#discussion_r318874373

liggitt replied:
We can't change the returned value of the subresource, and most subresources return a specific API type, which is why I thought we couldn't change it. 

The eviction subresource returns v1 Status however, so I *think* we can additionally accept a v1 Eviction type as long as we continue accepting v1beta1 Evictions as well (both v1 and v1beta1 would have to define conversions to the internal Eviction type).

We'd definitely need a test ensuring that both remained accepted.

mortent replied:
That makes sense. I've tried to look into this, and I'm struggling to see how we would handle this in the `PodExpansion` interface: https://github.com/kubernetes/kubernetes/blob/27c89b9aec73f66529a497910c460af2b25ab6dd/staging/src/k8s.io/client-go/kubernetes/typed/core/v1/pod_expansion.go#L33
I assume we can't add a separate function on the interface to handle the v1 version of the `Eviction`, so which type would we use here? Or should we be able to just keep the `v1beta1` type here and the `v1` version would be converted?

liggitt replied:
> I assume we can't add a separate function on the interface to handle the v1 version of the `Eviction`

That's exactly what I'd do. I'd probably add EvictV1 and EvictV1beta1, document EvictV1 to only apply to 1.21+, and mark `Evict` as deprecated to poke people to explicitly select a version (but not remove it any time soon).

mortent replied:
Updated the PR so we can accept both v1 Eviction types and v1beta1 Eviction types. This is done by adding additional `pods/evictionv1` and `pods/evictionv1beta1` endpoints and updating the `PodExpansion` with corresponding functions. It also adds test to verify that all endpoints works.

liggitt replied:
Hmm... I was hoping `pods/eviction` could accept either `policy/v1` or `policy/v1beta1` Evictions. I can believe that would not play nicely with the apiserver decoding stack, though. And currently, we can only advertise once accepted version in discovery.

@deads2k @lavalamp, any thoughts on what we should do here? To my knowledge, this is the only subresource of a GA endpoint that accepts a beta version. That's unfortunate to advertise forever and have no graduation path (and I worry a little about how the beta subresource type will interact with the "stop serving deprecated beta types automatically" [changes](https://github.com/kubernetes/kubernetes/pull/99362)).

lavalamp replied:
We never had anyone need to change the version of a subresource yet. The plan of record from way back in the day was to permit clients to use the Accept: header to request the version they want to see the result in. We should probably be emiting a warning to clients calling this without specifying an accept header, because such clients would start getting something different if we flip to having v1 as the default.

I think we must continue to accept v1beta1 for quite some time.

liggitt replied:
> I think we must continue to accept v1beta1 for quite some time.

No disagreement there... this is actually a pretty unique resource that is write-only (doesn't support GET and the result of the write is a v1.Status object), so we could conceivably accept both v1beta1 and v1 Eviction objects.

The question is whether we think that subresource accepting both versions is a good thing to aim for, especially if it is likely to involve fixing up bits in the discovery and endpoint handling stack that have single-version fixed assumptions.

lavalamp replied:
Oh this is write-only, that's great!

Sadly I don't see any option other than supporting both for some amount of time? I don't think it's OK to abruptly break old clients.

mortent replied:
So do we want to provide separate endpoints for the different version of the Eviction resource as we currently have in this PR or do we want to see if we can have one endpoint that can accept both the v1 and v1beta1 versions?

If we want to use one endpoint, it seems like that will require changes that are substantial enough that we can't make Eviction v1 in 1.21. So do we want to back out the changes to the Eviction resource from this PR and take just PDBs to GA in 1.21, or do we push everything to 1.22. 1.22 is the automatic deprecation deadline for PDBs?

liggitt replied:
> So do we want to provide separate endpoints for the different version of the Eviction resource as we currently have in this PR or do we want to see if we can have one endpoint that can accept both the v1 and v1beta1 versions?

One endpoint that accepts both would be my preference.

> If we want to use one endpoint, it seems like that will require changes that are substantial enough that we can't make Eviction v1 in 1.21. So do we want to back out the changes to the Eviction resource from this PR and take just PDBs to GA in 1.21, or do we push everything to 1.22.

I'm fine with moving out work on promoting eviction to v1 to 1.22.

> 1.22 is the automatic deprecation deadline for PDBs?

That's when their deprecation period starts. It ends in 1.25 and we stop serving them then.

soltysh replied:
> I'm fine with moving out work on promoting eviction to v1 to 1.22.

I was told we can't GA from a deprecated version, and we'd need to go through a 2nd beta, isn't that true?

liggitt replied:
If 1.22 adds Eviction v1 and marks Eviction v1beta1 deprecated, that seems reasonable. Eviction v1beta1 is an oddball since it's a subresource of a v1 API, so I expect we'll have interesting things to iron out there anyway.

mortent replied:
I have backed out the changes for updating Eviction to v1.

liggitt replied:
thanks, can you make sure this discussion is captured in an issue associated with the KEP, so we can make sure this happens in 1.22?

mortent replied:
Added a comment on the tracking issue: https://github.com/kubernetes/enhancements/issues/85#issuecomment-791591925

fedebongio commented:
/remove-sig api-machinery

mortent commented on pkg/registry/policy/poddisruptionbudget/strategy.go:
```diff
@@ -109,7 +111,8 @@ func (podDisruptionBudgetStatusStrategy) PrepareForUpdate(ctx context.Context, o
-	// TODO: Validate status updates.
-	return field.ErrorList{}
-	// return validation.ValidatePodDisruptionBudgetStatusUpdate(obj.(*policy.PodDisruptionBudget), old.(*policy.PodDisruptionBudget))
+	if requestInfo, found := genericapirequest.RequestInfoFrom(ctx); found && policyapiv1beta1.SchemeGroupVersion.Version == requestInfo.APIVersion {
```
This change is also in https://github.com/kubernetes/kubernetes/pull/98127, so I will those in here as soon as that PR is merged.

mortent commented:
/test pull-kubernetes-e2e-kind-ipv6

mortent commented:
/test pull-kubernetes-e2e-kind-ipv6
/test pull-kubernetes-node-e2e

fedebongio commented:
/remove-sig api-machinery

liggitt commented on pkg/registry/core/rest/storage_core.go:
```diff
@@ -111,7 +111,7 @@ func (c LegacyRESTStorageProvider) NewLegacyRESTStorage(restOptionsGetter generi
-	if policyGroupVersion := (schema.GroupVersion{Group: "policy", Version: "v1beta1"}); legacyscheme.Scheme.IsVersionRegistered(policyGroupVersion) {
+	if policyGroupVersion := (schema.GroupVersion{Group: "policy", Version: "v1"}); legacyscheme.Scheme.IsVersionRegistered(policyGroupVersion) {
 		var err error
 		podDisruptionClient, err = policyclient.NewForConfig(c.LoopbackClientConfig)
```
is a nil podDisruptionClient handled properly by eviction REST? if not, we expect policy/v1 to be enabled, so I would make this unconditional

liggitt commented on pkg/registry/policy/poddisruptionbudget/strategy.go:
```diff
@@ -114,8 +114,8 @@ func (podDisruptionBudgetStatusStrategy) ValidateUpdate(ctx context.Context, obj
 		apiVersion = schema.GroupVersion{
-			Group:   requestInfo.APIVersion,
-			Version: requestInfo.APIGroup,
+			Group:   requestInfo.APIGroup,
+			Version: requestInfo.APIVersion,
```
this is exciting... can you add a unit test that would have caught this?

mortent replied:
Yeah, I found this because of the unit test.

liggitt commented on pkg/apis/policy/v1/doc.go:
```diff
@@ -0,0 +1,25 @@
+// +k8s:defaulter-gen=TypeMeta
+// +k8s:defaulter-gen-input=../../../../vendor/k8s.io/api/policy/v1
+
+// Package policy is for any kind of policy object.  Suitable examples, even if
+// they aren't all here, are policyv1.PodDisruptionBudget, PodSecurityPolicy,
+// NetworkPolicy, etc.
```
suggest dropping the examples sentence, since PSP and NetworkPolicy won't end up here

liggitt commented on staging/src/k8s.io/api/policy/v1/types.go:
```diff
@@ -0,0 +1,147 @@
+	// +optional
+	MinAvailable *intstr.IntOrString `json:"minAvailable,omitempty" protobuf:"bytes,1,opt,name=minAvailable"`
+
+	// Label query over pods whose evictions are managed by the disruption
+	// budget.
```
Explicitly describe in documentation what `null` and `{}` selectors do. Also update the v1beta1 type documentation to indicate what `null` and `{}` selectors do.

It is extremely likely that some people will try to convert their v1beta1 PDB objects to v1 by simply changing the apiVersion, which will be unfortunate for them if the behavior of an empty selector changes.

mortent replied:
Updated.

liggitt commented on staging/src/k8s.io/client-go/listers/policy/v1/poddisruptionbudget_expansion.go:
```diff
@@ -0,0 +1,73 @@
+func (s *podDisruptionBudgetLister) GetPodPodDisruptionBudgets(pod *v1.Pod) ([]*policy.PodDisruptionBudget, error) {
+	var selector labels.Selector
+
+	if len(pod.Labels) == 0 {
+		return nil, fmt.Errorf("no PodDisruptionBudgets found for pod %v because it has no labels", pod.Name)
+	}
```
can't a v1 PDB with an empty selector match this?

liggitt replied:
actually, I'm pretty sure the v1beta1 version is incorrect here as well... a PDB selector could be expressed as a DoesNotExist selector, which could match pods without any labels

mortent replied:
Yeah, pods without labels don't actually work properly with PDBs, i.e. a pod without labels will not trigger a PDB reconcile. This is addressed in this PR by removing the check for empty labels in poddisruptionbudget_expansion.go. I also added additional integration tests to verify this behavior for both v1 and v1beta1.

liggitt commented on staging/src/k8s.io/client-go/listers/policy/v1/poddisruptionbudget_expansion.go:
```diff
@@ -0,0 +1,73 @@
+			continue
+		}
+
+		// Unlike the v1beta version, here we let an emty selector match everything.
```
```suggestion
		// Unlike the v1beta version, here we let an empty selector match everything.
```

liggitt commented on pkg/apis/policy/v1beta1/conversion.go:
```diff
@@ -0,0 +1,115 @@
+	pdbV1beta1Label = "pdb.kubernetes.io/deprecated-v1beta1-empty-selector-match"
+)
+
+func Convert_v1beta1_PodDisruptionBudget_To_policy_PodDisruptionBudget(in *v1beta1.PodDisruptionBudget, out *policy.PodDisruptionBudget, s conversion.Scope) error {
```
It's simpler to call the generated auto-conversion method first, which will handle all of the boilerplate, then tweak the selector specifically.

Rather than trying to strip the deprecated v1beta1 label specifically, I _think_ we can do two mechanical transformations in each direction by seeing if the input selector is identical to a specific "all" or "none" selector, and translating that into a new specific selector. Something like this:

```go

var (
	internalMatchAllSelector  = &metav1.LabelSelector{}
	internalMatchNoneSelector = &metav1.LabelSelector{
		MatchExpressions: []metav1.LabelSelectorRequirement{{Key: pdbV1beta1Label, Operator: metav1.LabelSelectorOpExists}},
	}

	v1beta1MatchNoneSelector = &metav1.LabelSelector{}
	v1beta1MatchAllSelector  = &metav1.LabelSelector{
		MatchExpressions: []metav1.LabelSelectorRequirement{{Key: pdbV1beta1Label, Operator: metav1.LabelSelectorOpDoesNotExist}},
	}
)

func Convert_v1beta1_PodDisruptionBudget_To_policy_PodDisruptionBudget(in *v1beta1.PodDisruptionBudget, out *policy.PodDisruptionBudget, s conversion.Scope) error {
	if err := autoConvert_v1beta1_PodDisruptionBudget_To_policy_PodDisruptionBudget(in, out, s); err != nil {
		return err
	}

	switch {
	case apiequality.Semantic.DeepEqual(in.Spec.Selector, v1beta1MatchNoneSelector):
		// If the v1beta1 version has a non-nil but empty selector, it should be
		// selecting no pods, even when used with the internal or v1 api. We
		// add a selector that is non-empty but will never match any pods.
		out.Spec.Selector = internalMatchNoneSelector.DeepCopy()
	case apiequality.Semantic.DeepEqual(in.Spec.Selector, v1beta1MatchAllSelector):
		// If the v1beta1 version has our v1beta1-specific "match-all" selector,
		// swap that out for a simpler empty "match-all" selector for v1
		out.Spec.Selector = internalMatchAllSelector.DeepCopy()
	}
	return nil
}

func Convert_policy_PodDisruptionBudget_To_v1beta1_PodDisruptionBudget(in *policy.PodDisruptionBudget, out *v1beta1.PodDisruptionBudget, s conversion.Scope) error {
	if err := autoConvert_policy_PodDisruptionBudget_To_v1beta1_PodDisruptionBudget(in, out, s); err != nil {
		return err
	}

	switch {
	case apiequality.Semantic.DeepEqual(in.Spec.Selector, internalMatchNoneSelector):
		// If the internal version has our v1beta1-specific "match-none" selector,
		// swap that out for a simpler empty "match-none" selector for v1beta1
		out.Spec.Selector = v1beta1MatchNoneSelector.DeepCopy()
	case apiequality.Semantic.DeepEqual(in.Spec.Selector, internalMatchAllSelector):
		// If the internal version has a non-nil but empty selector, we want it to
		// select all pods. We make sure this happens even with the v1beta1 api by
		// adding a non-empty selector that selects all pods.
		out.Spec.Selector = v1beta1MatchAllSelector.DeepCopy()
	}
	return nil
}
```

mortent replied:
I have updated the code to first use the auto-conversion and then fix the selector. 

But I think we might need the functionality to strip only the v1beta1 label, since there might be situations we need to convert and the v1beta1 label is not the only element in the selector. Since the selector in PDBs is mutable, I think it would be possible to read a resource as v1beta1, add an additional label to the PDB, and then write it back. If so, we can't rely on the equality checks.

This situation covered by this test: https://github.com/kubernetes/kubernetes/pull/99290/files#diff-ac28b0c739c72c12d59ca4e60b5d43c82c582bc87a005258b2412332d7d72ac3R89.

liggitt replied:
> Since the selector in PDBs is mutable, I think it would be possible to read a resource as v1beta1, add an additional label to the PDB, and then write it back.

I don't think that's an expected use case. Selectors are expected to be completely set by the writer... see https://github.com/kubernetes/kubernetes/issues/92913 and https://github.com/kubernetes/kubernetes/pull/93901

With the all-or-nothing conversion as proposed in #discussion_r589092943, what is the behavior in the following scenarios:

1. empty v1beta1 -> non-empty v1 -> empty v1beta1
    1. `kubectl apply` v1beta1 version of a PDB with an empty selector (`{}`)
    2. `kubectl apply` v1 version of the same PDB with some other selector (`{"matchLabels":{"foo":"bar"}}`) - expect only foo=bar via v1 and v1beta1 on read
    3. `kubectl apply` v1beta1 version of a PDB with an empty selector again (`{}`) - expect no trace of `foo=bar`

2. non-empty v1beta1 -> empty v1 -> non-empty v1beta1
    1. `kubectl apply` v1beta1 version of a PDB with a non-empty selector (`{"matchLabels":{"foo":"bar"}}`) - expect only foo=bar via v1 and v1beta1 on read
    2. `kubectl apply` v1 version of the same PDB with an empty selector (`{}`) - expect no trace of foo=bar
    3. `kubectl apply` v1beta1 version of a PDB with a non-empty selector again (`{"matchLabels":{"foo":"bar"}}`) - expect only foo=bar via v1 and v1beta1 on read

liggitt replied:
for this to behave as expected, I think the selector field has to have patchStrategy=replace

```
diff --git a/staging/src/k8s.io/api/policy/v1/types.go b/staging/src/k8s.io/api/policy/v1/types.go
index 42eab54b3c8..904413203ad 100644
--- a/staging/src/k8s.io/api/policy/v1/types.go
+++ b/staging/src/k8s.io/api/policy/v1/types.go
@@ -34,8 +34,9 @@ type PodDisruptionBudgetSpec struct {
 	// budget.
 	// A nil selector will match no pods, while an empty selector will select
 	// all pods within the namespace.
+	// +patchStrategy=replace
 	// +optional
-	Selector *metav1.LabelSelector `json:"selector,omitempty" protobuf:"bytes,2,opt,name=selector"`
+	Selector *metav1.LabelSelector `json:"selector,omitempty" patchStrategy:"replace" protobuf:"bytes,2,opt,name=selector"`
 
 	// An eviction is allowed if at most "maxUnavailable" pods selected by
 	// "selector" are unavailable after the eviction, i.e. even in absence of
diff --git a/staging/src/k8s.io/api/policy/v1beta1/types.go b/staging/src/k8s.io/api/policy/v1beta1/types.go
index 87c827d342b..4a1d2fee596 100644
--- a/staging/src/k8s.io/api/policy/v1beta1/types.go
+++ b/staging/src/k8s.io/api/policy/v1beta1/types.go
@@ -36,8 +36,9 @@ type PodDisruptionBudgetSpec struct {
 	// A nil selector selects no pods. An empty selector also selects no pods, which
 	// is different than the normal behavior for selectors where an empty selector
 	// does not restrict the selection space.
+	// +patchStrategy=replace
 	// +optional
-	Selector *metav1.LabelSelector `json:"selector,omitempty" protobuf:"bytes,2,opt,name=selector"`
+	Selector *metav1.LabelSelector `json:"selector,omitempty" patchStrategy:"replace" protobuf:"bytes,2,opt,name=selector"`
 
 	// An eviction is allowed if at most "maxUnavailable" pods selected by
 	// "selector" are unavailable after the eviction, i.e. even in absence of
```

@apelisse @jpbetz - does that match your expectations? is there something similar that needs to be specified for server-side apply to treat this field as atomic, or is that the default?

liggitt replied:
empty_v1beta1.yaml:
```yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
metadata:
  name: empty-v1beta1
spec:
  minAvailable: 2
  selector: {}
```

nonempty_v1.yaml:
```
apiVersion: policy/v1
kind: PodDisruptionBudget
metadata:
  name: empty-v1beta1
spec:
  minAvailable: 2
  selector:
    matchLabels:
      foo: bar
```


setting patchStrategy=replace and using a new kubectl aware of the go type change works as expected:

```
$ kubectl apply -f empty_v1beta1.yaml 
$ kubectl get -f empty_v1beta1.yaml  -o yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
...
spec:
  minAvailable: 2
  selector: {}

$ kubectl apply -f nonempty_v1.yaml 
$ kubectl get -f nonempty_v1.yaml -o yaml
apiVersion: policy/v1
kind: PodDisruptionBudget
...
spec:
  minAvailable: 2
  selector:
    matchLabels:
      foo: bar

$ kubectl apply -f empty_v1beta1.yaml 
$ kubectl get -f empty_v1beta1.yaml  -o yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
...
spec:
  minAvailable: 2
  selector: {}
```

using an old kubectl that doesn't have the go type change compiled in does not work:

```
$ kubectl-1.19 apply -f empty_v1beta1.yaml 
$ kubectl-1.19 get -f empty_v1beta1.yaml -o yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
...
spec:
  minAvailable: 2
  selector: {}

$ kubectl-1.19 apply -f nonempty_v1.yaml 
$ kubectl-1.19 get -f nonempty_v1.yaml -o yaml
apiVersion: policy/v1
kind: PodDisruptionBudget
...
spec:
  minAvailable: 2
  selector:
    matchExpressions:
    - key: pdb.kubernetes.io/deprecated-v1beta1-empty-selector-match
      operator: Exists
    matchLabels:
      foo: bar

$ kubectl-1.19 apply -f empty_v1beta1.yaml 
$ kubectl-1.19 get -f empty_v1beta1.yaml -o yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
...
spec:
  minAvailable: 2
  selector: {}
```

@apelisse @jpbetz - it looks like we can't change patchStrategy or set patchStrategy=replace on new go types and have old kubectl clients client-side apply correctly?

liggitt replied:
kubectl 1.19 sends `"Content-Type: application/merge-patch+json"` when applying to `policy/v1`:
```json
{
    "metadata": {
        "annotations":{
            "kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"policy/v1\",\"kind\":\"PodDisruptionBudget\",\"metadata\":{\"annotations\":{},\"name\":\"empty-v1beta1\",\"namespace\":\"default\"},\"spec\":{\"minAvailable\":2,\"selector\":{\"matchLabels\":{\"foo\":\"bar\"}}}}\n"
        }
    },
    "spec":{
        "selector":{
            "matchLabels":{"foo":"bar"}
        }
    }
}
```

kubectl build from this branch sends `"Content-Type: application/strategic-merge-patch+json"` when applying to `policy/v1`, and works properly:
```json
{
    "metadata": {
        "annotations": {
            "kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"policy/v1\",\"kind\":\"PodDisruptionBudget\",\"metadata\":{\"annotations\":{},\"name\":\"empty-v1beta1\",\"namespace\":\"default\"},\"spec\":{\"minAvailable\":2,\"selector\":{\"matchLabels\":{\"foo\":\"bar\"}}}}\n"
        }
    },
    "spec":{
        "selector":{
            "matchLabels":{"foo":"bar"}
        }
    }
}
```

liggitt replied:
I thought kubectl would read patch info from openapi and send strategic merge patches even on types it did not have compiled in

liggitt replied:
fortunately, both old and new kubectl's behave like a champ with `apply --server-side`

liggitt replied:
looks like kubectl doesn't send strategic merge patches for unknown types, even if the openapi schema indicates they are supported:

https://github.com/kubernetes/kubernetes/blob/4fccba9e068f31a562f753caf4b37959b4ef829a/staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go#L120-L166

I'll open a separate issue to track that, but a fix there will be too late for this to take advantage of. Given that, I think you approach of stripping the deprecated beta label would work well as a default case for the v1beta1->internal conversion (if we don't exactly match one of the well-known selectors). I think we also need to do that check in v1->internal conversion so that when we switch etcd storage to v1 (and never hit this v1beta1 conversion code), we'll catch selectors like that.

Messing around with that in the last commit of https://github.com/liggitt/kubernetes/commits/PromotePDBsToGA seemed to work well.

seh replied:
I thought this rang a bell with a problem in _kubecfg_ from two years ago. See https://github.com/bitnami/kubecfg/issues/259#issuecomment-522054301. The related discussion in the "sig-api-machinery" Slack channel starts [here](https://kubernetes.slack.com/archives/C0EG7JC6T/p1565969006291300).

mortent replied:
Updated based on the suggested solution at https://github.com/liggitt/kubernetes/commits/PromotePDBsToGA

liggitt commented:
along with this, we'll want to:
1. merge a change to switch the existing e2e test to use the v1 API
2. follow-up with a change to switch that e2e test to be a conformance test

can you add a commit to this PR to switch the e2e test to use v1, and open a follow-up PR to promote the test to conformance (so we can make sure `pull-kubernetes-conformance-kind-ga-only-parallel` passes on it)?

mortent commented:
/test pull-kubernetes-e2e-kind

mortent commented:
@liggitt I have addressed/responded to the comments, and added a separate commit to move all e2e tests to use the v1 api.
I have created a PR to promote e2e tests to Conformance: https://github.com/kubernetes/kubernetes/pull/84740

liggitt commented on staging/src/k8s.io/client-go/listers/policy/v1/poddisruptionbudget_expansion.go:
```diff
@@ -0,0 +1,69 @@
+type PodDisruptionBudgetNamespaceListerExpansion interface{}
+
+// GetPodPodDisruptionBudgets returns a list of PodDisruptionBudgets matching a pod.
+func (s *podDisruptionBudgetLister) GetPodPodDisruptionBudgets(pod *v1.Pod) ([]*policy.PodDisruptionBudget, error) {
```
should this be fixed in staging/src/k8s.io/client-go/listers/policy/v1beta1/poddisruptionbudget_expansion.go as well?

```go
if len(pod.Labels) == 0 {
	return nil, fmt.Errorf("no PodDisruptionBudgets found for pod %v because it has no labels", pod.Name)
}
```

mortent replied:
Yeah, I have updated it.

liggitt reviewed (commented):
> Should I update https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L787 to list PDB as GA in this PR? I suspect this is why the conformance tests for PDBs fail when running with GA only.

Yes

mortent commented on test/e2e/apps/disruption.go:
```diff
@@ -27,6 +27,7 @@ import (
 	appsv1 "k8s.io/api/apps/v1"
 	v1 "k8s.io/api/core/v1"
+	policyv1 "k8s.io/api/policy/v1"
 	policyv1beta1 "k8s.io/api/policy/v1beta1"
```
So I think the test failures in #84740 is primarily because the feature flag wasn't updated, so the disruption controller isn't running in those tests. I have updated this PR to address that and will also update #84740.
Another issue here is that some of the pdb e2e tests use the eviction API, which of course is not GA. So I will look at the tests being promoted to conformance in #84740 and make sure those doesn't rely on Eviction. But this means we still need to import `policyv1beta1 "k8s.io/api/policy/v1beta1"` here.

liggitt replied:
> Another issue here is that some of the pdb e2e tests use the eviction API, which of course is not GA. So I will look at the tests being promoted to conformance in #84740 and make sure those doesn't rely on Eviction. But this means we still need to import `policyv1beta1 "k8s.io/api/policy/v1beta1"` here.

ah, that makes sense for the import. I think we'll want to continue exercising the subresource in the test, there's little use for the PDB API otherwise

mortent commented:
/test pull-kubernetes-e2e-kind

mortent commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

mortent commented:
@liggitt The PR is updated and all tests are passing. The conformance tests also pass the `pull-kubernetes-conformance-kind-ga-only-parallel` job on https://github.com/kubernetes/kubernetes/pull/84740. 
I couldn't promote all tests I would have liked to conformance due to the dependencies on the Eviction API. We should make sure that is addressed when we GA Eviction.

liggitt commented:
/lgtm
/approve

mortent committed:
generated

mortent committed:
Update e2e tests to use the policy v1 api

liggitt commented:
/lgtm

mortent commented:
/test pull-kubernetes-e2e-kind

liggitt commented:
flake was https://github.com/kubernetes/kubernetes/issues/99979

mortent commented:
/test pull-kubernetes-e2e-kind

mortent commented:
Flake seems related to https://github.com/kubernetes/kubernetes/issues/96337

liggitt commented:
/retest

liggitt commented:
flakes were #99911 and https://github.com/kubernetes/kubernetes/pull/94564

mortent commented:
/test pull-kubernetes-integration

liggitt commented:
integration flake is https://github.com/kubernetes/kubernetes/pull/98116#issuecomment-794256319

wilsonehusin commented:
hey folks, 1.21 Release Notes team here -- want to check in and disambiguate this line:

> The only functional change is that an empty selector (`{}`) written to a policy/v1 PodDisruptionBudget now selects all pods in the namespace.

does this behavior warrant this note to be in [`ACTION REQUIRED` / Urgent Upgrade Notes](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.21/release-notes/release-notes-draft.md#urgent-upgrade-notes)? I'm not too familiar with how `PodDisruptionBudget` works when given empty selector, and if this potentially creates side effect to end users on `v1beta1` -> `v1`, then we might want to surface this note 😄 

cc @mortent @liggitt

liggitt commented:
no, there are no side effects to existing v1beta1 consumers
